### PR TITLE
Magpie upgrade strike II

### PR DESCRIPTION
**Pull request by huard**
Strike II of this original PR https://github.com/bird-house/birdhouse-deploy/pull/107.

Matching notebook fix https://github.com/Ouranosinc/pavics-sdi/pull/218

Performed test upgrade on staging (Medus) using prod (Boreas) Magpie DB, everything went well and Jenkins passed (http://jenkins.ouranos.ca/job/ouranos-staging/job/medus.ouranos.ca/80/parameters/).  This Jenkins build uses the corresponding branch in https://github.com/Ouranosinc/pavics-sdi/pull/218 and with `TEST_MAGPIE_AUTH` enabled.

Manual upgrade migration procedure:
1. Save `/data/magpie_persist` folder from prod `pavics.ouranos.ca`: `cd /data; tar czf magpie_persist.prod.tgz magpie_persist`
2. scp `magpie_persist.prod.tgz` to `medus`
3. login to `medus`
4. `cd /path/to/birdhouse-deploy/birdhouse`
3. `./pavics-compose.sh down`
4. `git checkout master`
5. `cd /data`
2. `rm -rf magpie_persist`
3. `tar xzf magpie_persist.prod.tgz`  # restore Magpie DB with prod version
4. `cd /path/to/birdhouse-deploy/birdhouse`
5. `./pavics-compose.sh up -d`
3. Update `env.local` `MAGPIE_ADMIN_PASSWORD` with prod passwd for Twitcher to be able to access Magpie since we juste restore the Magpie DB from prod
4. `./pavics-compose.sh restart twitcher`  # for Twitcher to get new Magpie admin passwd 
4. Baseline working state: trigger Jenkins test suite, ensure all pass except `pavics_thredds.ipynb` that requires new Magpie
5. Baseline working state: view existing services permissions on group Anonymous (https://medus.ouranos.ca/magpie/ui/groups/anonymous/default)
6. `git checkout restore-previous-broken-magpie-upgrade-so-we-can-work-on-a-fix`  # This current branch
7. `./pavics-compose.sh up -d`  # upgrade to new Magpie
8. `docker logs magpie`: check no DB migration error
9. Trigger Jenkins test suite again

### CODE REVIEW

**Comment by huard** [comment]
ATM: `MAGPIE_CLI_IMAGE="pavics/magpie:3.2.0"` is this ok ? Does it need to be set to 3.8.0 or 1.7.3 ?

**Comment by MatProv** [comment]
run tests

File: birdhouse/docker-compose.yml [hunk 1]
```diff
@@ -297,6 +297,7 @@ services:
       MAGPIE_PROVIDERS_CONFIG_PATH: "/opt/local/src/magpie/config/providers"
       MAGPIE_PERMISSIONS_CONFIG_PATH: "/opt/local/src/magpie/config/permissions"
       MAGPIE_POSTGRES_HOST: postgres-magpie
+      MAGPIE_PORT: ${MAGPIE_PORT}
```
Comment (tlvu) [inline birdhouse/docker-compose.yml:300]: @dbyrns This was not needed with previous Magpie 3.5.1, so probably not be needed now either.  Anyways, there's nothing in `default.env` and `env.local.example` with this var so what would be the value of this `MAGPIE_PORT`?

**Review by tlvu — COMMENTED** [review]
Start working on this, let's try to get the automation pipeline working first before attempting an upgrade migration test.

There is a corresponding pavics-sdi branch https://github.com/Ouranosinc/pavics-sdi/tree/restore-change-for-new-broken-magpie-so-we-can-test-magpie-fix to use with Jenkins.

**Comment by tlvu** [comment]
Looks like Magpie could not start due to
```
[2021-04-07 02:45:21,113] ERROR      [MainThread][magpie.utils] Constant could not be found: MAGPIE_PORT
```
See https://daccs-jenkins.crim.ca/job/DACCS-iac/275/artifact/docker-compose-by-service-host-140-8.rdext.crim.ca.log

As I suspected, we went from 3.5.1 to 3.8.0.  So there might be additional config change.  Just bumping the Magpie docker tag is not all there is to it.

**Comment by tlvu** [comment]
> Looks like Magpie could not start due to
> 
> ```
> [2021-04-07 02:45:21,113] ERROR      [MainThread][magpie.utils] Constant could not be found: MAGPIE_PORT
> ```
> 
> See https://daccs-jenkins.crim.ca/job/DACCS-iac/275/artifact/docker-compose-by-service-host-140-8.rdext.crim.ca.log
> 
> As I suspected, we went from 3.5.1 to 3.8.0. So there might be additional config change. Just bumping the Magpie docker tag is not all there is to it.

So adding `MAGPIE_PORT` env var to the magpie container got it to start.  But I am confused, what value is `MAGPIE_PORT` since it is not defined anywhere in the stack.

File: birdhouse/docker-compose.yml (hunk 1 above)
Comment (dbyrns) [reply birdhouse/docker-compose.yml]: The default value has been dropped in Magpie, but you're right no need to add a variable for that. I set it directly as for others birds.

**Comment by tlvu** [comment]
run tests

**Comment by dbyrns** [comment]
> As I suspected, we went from 3.5.1 to 3.8.0. So there might be additional config change. Just bumping the Magpie docker tag is not all there is to it.

To be discussed this afternoon. I bite the bullet to experience what it is to test an update and this is not something funny.
We have to test : 
 https://github.com/bird-house/birdhouse-deploy/tree/restore-previous-broken-magpie-upgrade-so-we-can-work-on-a-fix
with : 
https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/tree/magpie-auth-fixes
and : 
https://github.com/Ouranosinc/pavics-sdi/tree/restore-change-for-new-broken-magpie-so-we-can-test-magpie-fix
and also our private config repo which is not the same as yours.

So 4 repos and I don't even take into account production data. So when a test goes well for someone, it can still goes bad elsewhere. We will think about how we could streamline all of that even more so we can increase our confidence in tests.

**Comment by dbyrns** [comment]
> run tests

@tlvu, indeed the tests cannot passed here because it used both master branch of PAVICS-e2e-workflow-tests and pavics-sdi. That's why Francis needs to test with a custom iac. We (CRIM and Ouranos) will need to find a way to test automatically any branches of all involded repos if we want a full tracability and reproductibility that cannot be provided right now.

**Comment by tlvu** [comment]
> > run tests
> 
> @tlvu, indeed the tests cannot passed here because it used both master branch of PAVICS-e2e-workflow-tests and pavics-sdi. That's why Francis needs to test with a custom iac. 

Yes I am fully aware of this.  I started a custom iac that do not destroy at end https://daccs-jenkins.crim.ca/job/DACCS-iac/276/, then manually started the Jenkins with proper pavics-sdi branch on that instance https://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/256/parameters/ (it passed by the way, with `notebooks-auth/test_thredds.ipynb`, youpiiii).

https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/tree/magpie-auth-fixes was not used since that branch is mostly to delete duplicate cells and add more explanation to the `notebooks-auth/test_thredds.ipynb` notebook.  I was going to try this branch once I see the result without first.

So sorry for doing a `run tests` here.  I understand a regular `run tests` would not have worked.  I was doing a tutorial to @huard and other collegues at Ouranos on our mutual Jenkins infrastructure so I picked this PR since it was already open, sorry my bad.

> We (CRIM and Ouranos) will need to find a way to test automatically any branches of all involded repos if we want a full tracability and reproductibility that cannot be provided right now.

Agreed.

**Comment by dbyrns** [comment]
> it passed by the way, with notebooks-auth/test_thredds.ipynb, youpiiii

Are you saying that we could go ahead and finally merge this PR? (No quick merge this time, just asking)

**Comment by tlvu** [comment]
> > it passed by the way, with notebooks-auth/test_thredds.ipynb, youpiiii
> 
> Are you saying that we could go ahead and finally merge this PR? (No quick merge this time, just asking)

After I try the newer commits in this PR together with the https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/tree/magpie-auth-fixes branch and then perform a manual upgrade migration as described here https://github.com/Ouranosinc/Magpie/issues/403 (Section "To Reproduce" at the end of the PR description).

Only then we can merge because, with autodeploy, a merge is equivalent to a go-live so I am really careful.

**Comment by dbyrns** [comment]
The jenkins bot comment is really cool!! Thanks @MatProv

**Comment by tlvu** [comment]
> The jenkins bot comment is really cool!! Thanks @MatProv

Lots clearer what has been tested (the combination of the different changing pieces) !!!  Super @MatProv 

So for this to work, I assume we have to put the PR when requesting the build manually?  A "run tests" in the comment will not be able to specify the different combinations?  Not suggesting to fix "run tests" either. I am fine with using Jenkins interface.

Can I just suggest again to put the httpS link to jenkins because for us outside of your CRIM VPN, http do not work and we always have to manually change the link to httpS for each and every click.  It's getting very annoying.

**Comment by dbyrns** [comment]
Ok we got the good branches and params. #348 is as good as we can regarding Magpie. Failure is due to cells output mismatch unrelated to this PR.
For upcoming PRs we sadly cannot expect better result than this for now. Hopefully notebooks will get fix.

**Comment by MatProv** [comment]
@tlvu Build 348's magpie tests pass - the failure only indicates notebooks diff that fail. But there are no more AuthenticationFailure.

> So for this to work, I assume we have to put the PR when requesting the build manually?

Yes exactly, the issue we had before with the builds triggered via jenkins is that it didn't updated the PR (neither the build status or comments). This modification aims to link jenkins build to github PRs. Any type of github trigger (on commit, on comment) triggers the "basic" build status update without adding a comment. Might easily be modify later on, but the main goal here with pipeline triggered via jenkins (for custom params and checkouts) is reached.

**Comment by MatProv** [comment]
run tests

**Comment by tlvu** [comment]
> Failure is due to cells output mismatch unrelated to this PR.

Most likely notebook update for the newer Jupyter env that has not been merged with the pavics-sdi branch.  I can take care of this as I know what has changed to make it work for the Jupyter env.

**Comment by MatProv** [comment]
run tests

**Comment by MatProv** [comment]
run tests

**Comment by MatProv** [comment]
run tests

**Comment by MatProv** [comment]
run tests

**Comment by tlvu** [comment]
Performed test upgrade on staging (Medus) using prod (Boreas) Magpie DB, everything went well and Jenkins passed (http://jenkins.ouranos.ca/job/ouranos-staging/job/medus.ouranos.ca/80/parameters/).  This Jenkins build uses the corresponding branch in https://github.com/Ouranosinc/pavics-sdi/pull/218 and with `TEST_MAGPIE_AUTH` enabled.

**Comment by tlvu** [comment]
> # DACCS-iac Pipeline Results
> Build Number : 361
> Build URL : http://daccs-jenkins.crim.ca:80/job/DACCS-iac/361/
> Result : failure
> 
> BIRDHOUSE_DEPLOY_BRANCH : restore-previous-broken-magpie-upgrade-so-we-can-work-on-a-fix
> DACCS_CONFIGS_BRANCH : master
> PAVICS_E2E_WORKFLOW_TESTS_BRANCH : master
> PAVICS_SDI_BRANCH : restore-change-for-new-broken-magpie-so-we-can-test-magpie-fix
> 
> DESTROY_INFRA_ON_EXIT : true

pavics_thredds.ipynb: HTTPError: 401 Client Error: Unauthorized for url: https://host-140-46.rdext.crim.ca/magpie/signin with the authtest/authtest1234 user.  Weird, user creation error at the beginning?  This also pas on my Jenkins run against Medus.

subset.ipynb: failed with some ProcessFailed but this works on my Jenkins run.

The pipeline seems a bit unstable.

**Review by tlvu — APPROVED** [review]

**Pull request merged**